### ISSUE TITLE
Windows authentication request for back up tool

### ISSUE THREAD

**Issue by psrinivasa2** [wiltondb/wiltondb#34]
Hi Alex @staticlibs,
Can the windows authentication request be added back up tool wdb_backup?

also we we try to take backup for our database, backup failed with below error

![image](https://github.com/user-attachments/assets/1a36312b-27a4-4050-bf47-7681b3545b6f)

I would like to remind you about the same problem which we faced earlier and you added the fix for the same, you may need to add same fix in back up tool. could you please check on it?

![image](https://github.com/user-attachments/assets/efd49f27-b1fe-4df2-8990-510f6b55440e)

**Comment by staticlibs** [comment]
@psrinivasa2 

> Hi Alex @staticlibs, Can the windows authentication request be added for back up tool wdb_backup?

No, this is not planned. `pg_dump` works over Postgres protocol, and using SSPI with Postgres connections is much less convenient than with TDS ones (some details are in [this comment](https://github.com/wiltondb/wiltondb/issues/18#issuecomment-2263366339)).

Some other Postgres auth feature can be used to handle passwords in this case, for example [a password file](https://www.postgresql.org/docs/15/libpq-pgpass.html).

> 
> also when we try to take backup for our database, backup failed with below error
> 
> I would like to remind you about the same problem which we faced earlier and you added the fix for the same, you may need to add same fix in back up tool. could you please check on it?
> 

Thanks for the details! I also think that the problem is related to #24. But for some reason I cannot reproduce it with `pg_dump`. Tried both `binary(8)` and `rowversion` columns and `pg_dump` (called from `wdb_backup`) dumps it successfully for me.

Would you mind simplifying your table (the one with `scn` column) DDL to reproduce the problem with minimal number of columns and data records? Alternatively, if you run `wdb_backup` on debug version of WiltonDB MSI and collect the backtrace for this error - it also may be helpful.

**Comment by staticlibs** [comment]
The problem happens [here](https://github.com/wiltondb/postgresql_modified_for_babelfish/blob/8b86250fbded7c6134c754ef6ba760501b9d60ef/src/backend/commands/copy.c#L780), when `rowversion` column is requested to be dumped by `pg_dump`. There seems to be a discrepancy in detecting when column is `rowversion` and needs [to be excluded](https://github.com/wiltondb/postgresql_modified_for_babelfish/blob/8b86250fbded7c6134c754ef6ba760501b9d60ef/src/bin/pg_dump/pg_dump.c#L18415) by `pg_dump`. 

Would you mind checking the registered type for the problematic column with the following query in PgAdmin (replacing table and column names):

```sql
select att.attname, typ.typname, typ.oid from pg_attribute att
join pg_type typ on typ.oid = att.atttypid
where att.attrelid = 'test1_dbo.tab1'::regclass::oid
and att.attname = 'col1'
```

And the OIDs for the registered types:

```sql
select typname, oid from pg_type
where typname in ('binary', 'rowversion', 'timestamp')
```

Can you also try to drop and re-create the table on the latest MSI version and check the type again?

**Comment by Aravind2601** [comment]
Hi @staticlibs ,
I have tried drop and recreate table with latest MSI below is the result of the query before and after .
wilton - 10.15.0.0
![image](https://github.com/user-attachments/assets/09b50486-6b8f-4c65-bf40-7844b26307b9)
wilton-11.15.1.0
![image](https://github.com/user-attachments/assets/72f825e5-21a2-468d-9f98-cd1307891cb6)
OIDs for the registered types:
![image](https://github.com/user-attachments/assets/96cd75cf-c16a-49b3-a5ba-3cec14d750b3)

**Comment by Aravind2601** [comment]
Hi @staticlibs ,
we have tried password file option for wilton backup tool but it did not credentials from there.
Below is the path i have the passconfig file.
![image](https://github.com/user-attachments/assets/d258af5e-370d-4258-b4cd-9225582a0b02)
Below is the details of pass file.
![image](https://github.com/user-attachments/assets/fc06fbbb-445c-43f8-9841-c216a347ae11)

Any idea of what we missing there.

**Comment by staticlibs** [comment]
@Aravind2601 , thanks for details! Assuming `scn` column must be `binary(8)`, that means that column type was confused with `timestamp` at table creation time. If you can reproduce this - can you share a standalone (simplified) T-SQL DDL that creates the table - so I can reproduce this locally?

PS: I will comment on pgpass file separately in a few hours.

**Comment by Aravind2601** [comment]
@staticlibs  I have sent in mail.

**Comment by staticlibs** [comment]
@Aravind2601 , on `pgpass` file, please provide more details on the error you are getting. Is it something like the following, or a different one:

```
> psql -h 127.0.0.1 -p 5432 -U wilton
psql: error: connection to server at "127.0.0.1", port 5432 failed: FATAL:  password authentication failed for user "wilton"
password retrieved from file "C:\Users\alexk\AppData\Roaming/postgresql/pgpass.conf"
connection to server at "127.0.0.1", port 5432 failed: FATAL:  password authentication failed for user "wilton"
password retrieved from file "C:\Users\alexk\AppData\Roaming/postgresql/pgpass.conf"
```

I suppose your problem is in `localhost` part, depending how local address is resolved. I suggest to use something like this:

```
*:*:*:wilton:wilton
```

**Comment by staticlibs** [comment]
@Aravind2601 , on `rowversion/timestamp` problem - thanks for providing the DDL! The problem is reproducible with the following table:

```tsql
create table tab1(col1 int, col2 timestamp)
```

Curiously enough, it is NOT reproducible with:

```tsql
create table tab1(col1 int, col2 rowversion)
```

despite `rowversion` and `timestamp` being synonyms. It appeared that typename is returned incorrectly quoted for T-SQL `timestamp` columns and this breaks `COPY` command called by `pg_dump`. I was unable to quickly find a root cause on server, instead added a workaround to `pg_dump`.  Updated `wdb_backup` to [version 1.0.5](https://github.com/wiltondb/wdb_backup/releases/tag/1.0.5) including fixed `pg_dump.exe`. Please try this version for GUI backups, and also use `pg_dump.exe` from there (instead of one in `C:\Program Files\WiltonDB Software\wiltondb3.3\bin`) for command-line backup calls.

I will keep this issue open for a proper fix in DB backend and potential upstreaming to Babelfish.

**Comment by bill-ramos-rmoswi** [comment]
Given the way Babelfish works with PostgreSQL and MVCC transactions under the cover, I recommend to all customers to remove timestamp and rowversion columns from their tables. Under the covers, Babelfish implements the columns using the "hidden" system columns in pg_attributes. The actually show down transactions when present. The values are completely useless when you transfer them with pg_dump / restore. So, delete them!

Here are the specifics from Claude.ai you can follow up on.

PostgreSQL system columns, including ctid, play important roles in transaction management and other database functions. Let's explore some key system columns:

1. ctid (Cursor Tuple ID):

- Format: (block_number, tuple_index_within_block)
- Functions:
  - Unique identifier for each row version within a table
  - Used for row-level locking
  - Enables efficient updates by creating new row versions
- Transaction management: Helps implement Multi-Version Concurrency Control (MVCC)


2. xmin (Transaction ID that inserted the row):

- Indicates which transaction created this row version
- Used for determining row visibility to concurrent transactions


3. xmax (Transaction ID that deleted or updated the row):

- Null if row is still valid
- Set to the transaction ID that deleted or updated this row version
- Used for determining row visibility and cleanup of old versions


4. cmin (Command identifier within the inserting transaction):

- Helps order operations within a single transaction


5. cmax (Command identifier within the deleting transaction):

- Similar to cmin, but for the deleting transaction


6. tableoid:

- Contains the OID of the table containing this row
- Useful in inheritance hierarchies to identify which table a row belongs to


7. oid (Object Identifier):

- Deprecated in modern PostgreSQL versions
- Previously used as a unique row identifier


These system columns contribute to PostgreSQL's MVCC implementation, enabling:

- Concurrent access without read locks
- Consistent snapshots of the database
- Efficient implementation of transactions and isolation levels

**Comment by Aravind2601** [comment]
@staticlibs,  thanks for the fix in version 1.0.5 . its working fine with this exe. 
Regarding the password file pgadmin is taking the credentials from there. But backup tool is not taking those values it still ask for password . Could you please help use on this?

**Comment by staticlibs** [comment]
@Aravind2601 , yes, I will add support for pgpass file to wdb_backup and wdb_config.

**Comment by psrinivasa2** [comment]
Hi Alex@staticlibs ,
Our plan is to use OS user in all the place, as you mentioned in different ticket, we will generate random unknown password for wilton user. Please advise us in terms of OS user in back tool, configuration tool and other required places

**Comment by staticlibs** [comment]
@psrinivasa2 , it may be problematic to use OS users with current version. With `sspi` auth set up in `pg_hba.conf` only Postgres logins will work. TDS logins currently can only be done when password (`md5`) auth is used in `pg_hba.conf`. If you intend to use SSPI logins on both TDS and Postgres ports it makes sense to add support for `sspi` in `pg_hba.conf` and also for `pg_ident.conf` maps for TDS logins (to have the same setup effective for both Postgres and TDS ports). This is possible, but this is not a trivial task.

**Comment by staticlibs** [comment]
Just for the record, submitted [bug report upstream](https://github.com/babelfish-for-postgresql/babelfish_extensions/issues/2902) about the `timestamp` problem with `pg_dump`. Modified `pg_dump` from `wdb_backup-1.0.5` (with a workaround to this problem) is going to be included into upcoming MSI installer update.

**Comment by staticlibs** [comment]
@Aravind2601 , about the "Cannot determine schema name" error: I've filed #46 for it and updated `wdb_backup` with the fix.

**Comment by Aravind2601** [comment]
@staticlibs thanks alex, will check

**Comment by Aravind2601** [comment]
@staticlibs, we checked with above latest backup tool but the restore was failed. I have shared the log details in mail. Please check and let us know if you got any details.

**Comment by staticlibs** [comment]
@Aravind2601 

Thanks for the logs! First 2 errors there are related to schema name rewriting (for DB renaming during restore) in views/procedures. It seems that SQL tokenizer library I've used does not support some of the naming used in T-SQL (like `$` characters in names). I will follow-up on this.

I wonder if you can check the details about `DB3.txt` on your side? DB rename is performed successfully there, but then it fails on constraint violation. I wonder if you can check what is this constraint doing and if it is actually being violated by the restored records?

**Comment by Aravind2601** [comment]
@staticlibs Any update on above requests

**Comment by staticlibs** [comment]
@Aravind2601 

Working on `DB1` and `DB2` problems (different problems in the same area), and `DB3` behaviour is assumed to work as expected (unless something is wrong with constraint handling - cannot tell without seeing the constraint and corresponding matching record).

**Comment by Aravind2601** [comment]
@staticlibs  we will check the DB3 from our end and then let you know Alex, you can analyze and help to resolve the issue in DB1 and DB2.

**Comment by staticlibs** [comment]
@Aravind2601

I've filed #55 to track fixes for DB1 and DB2 - please see the details (and the updated version of `wdb_backup`) there.

**Comment by staticlibs** [comment]
The fix is included in [update 12.16.1](https://github.com/wiltondb/wiltondb/releases/tag/3-lts-12-16-1).
